pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/autodetect.go

```diff
@@ -0,0 +1,48 @@
+package tpmutil
+
+import (
+	"errors"
+	"io/ioutil"
+	"path"
+	"regexp"
+
+	"github.com/google/go-tpm/tpm2"
+)
+
+const baseTPMDir = "/dev"
+
+var validTPMNames = []*regexp.Regexp{
+	regexp.MustCompile(`tpmrm\d+$`),
+	regexp.MustCompile(`tpm\d+$`),
+}
+
+func AutoDetectTPMPath() (string, error) {
+	files, err := ioutil.ReadDir(baseTPMDir)
+	if err != nil {
+		return "", err
+	}
+
+	for _, validExp := range validTPMNames {
+		for _, f := range files {
+			if validExp.MatchString(f.Name()) {
+				candidateTPM := path.Join(baseTPMDir, f.Name())
+				if isValidTPM(candidateTPM) {
+					return candidateTPM, nil
+				}
+			}
+		}
+	}
+
+	return "", errors.New("unable to autodetect TPM")
+}
+
+// VerifyTPM verifies that the given path belongs to a functional TPM 2.0
+func isValidTPM(path string) bool {
+	rwc, err := tpm2.OpenTPM(path)
+	if err != nil {
+		return false
```

Comment:
> There can be "other" things that cause this too, right? For example, I think TPM is only accessible by root, so if agent is not root this can fail? And then we will report "unable to autodetect TPM".

That's a good point.

> How likely do you think it is that there is more than one device that matches this pattern?

The only reason I can think of for having more than one matching entry is in case the machine have more than one TPM chip.

>  how likely is it that there's a device that matches this pattern that is not a TPM?

I think this is unlikely

> I wonder if it is enough for us to just say "if we find exactly one device that matches this pattern, we will try to use it. if we find zero or more than one, we will ask you to configure it"

I agree! That's a better approach. I'll update the code to match that behavior.